R/xgb_binning.R

```diff
@@ -0,0 +1,250 @@
+#'
+
+step_xgb_binning <-
+  function(recipe,
+           ...,
+           role = NA,
+           trained = FALSE,
+           outcome = NULL,
+           rules = NULL,
+           prefix = "xgb_bin",
+           skip = FALSE,
+           id = rand_id("xgb_binning")) {
+    if (is.null(outcome))
+      stop("`outcome` should select at least one column.", call. = FALSE)
+
+    # recipes_pkg_check("xgb_binninb") - TO DO CHECKS
+
+    add_step(
+      recipe,
+      step_xgb_binning_new(
+        terms = ellipse_check(...),
+        role = role,
+        trained = trained,
+        outcome = outcome,
+        rules = rules,
+        prefix = prefix,
+        skip = skip,
+        id = id
+      )
+    )
+  }
+
+step_xgb_binning_new <-
+  function(terms, role, trained, outcome, rules,
+           prefix, skip, id) {
+    step(
+      subclass = "xgb_binning",
+      terms = terms,
+      role = role,
+      trained = trained,
+      outcome = outcome,
+      rules = rules,
+      prefix = prefix,
+      skip = skip,
+      id = id
+    )
+  }
+
+run_xgboost <- function(.train, .test, .max_bin){
+  xgboost::xgb.train(
+    params = list(
+      eta = 0.3,
+      max_bin = .max_bin,
+      max_depth = 1,
+      min_child_weight = 5
+    ),
+    nrounds = 100,
+    data = .train,
+    watchlist = list(
+      train = .train,
+      test  = .test
+    ),
+    tree_method = "hist",
+    early_stopping_rounds = 10,
+    objective = "binary:logistic",
+    eval_metric = "auc",
+    verbose = 0
+  )
+}
+
+xgb_binning <- function(df, target, variable){
+
+  levels <- sort(unique(df[[target]]))
+
+  if(length(levels) != 2){
+    stop("Target variable does not have exactly two levels.", call. = FALSE)
+  }
+
+  if(any(is.na((levels)))){
+    stop("Target variable contains missing values.", call. = FALSE)
+  }
+
+  split <- rsample::initial_split(df, prop = 0.8, strata = target)
+  train <- rsample::training(split)
+  test  <- rsample::testing(split)
+
+  xgb_train <- xgboost::xgb.DMatrix(
+    data = as.matrix(train[, variable]),
+    label = ifelse(train[[target]] == levels[[1]], 0, 1)
+  )
+
+  xgb_test <- xgboost::xgb.DMatrix(
+    data = as.matrix(test[, variable]),
+    label = ifelse(test[[target]] == levels[[1]], 0, 1)
+  )
+
+  xgb_runs <- purrr::map(c(10, 20, 50), function(x) run_xgboost(xgb_train, xgb_test, x))
+
+  xgb_runs_results <- rbind(
+    xgb_runs[[1]]$evaluation_log[xgb_runs[[1]]$best_iteration, ],
+    xgb_runs[[2]]$evaluation_log[xgb_runs[[2]]$best_iteration, ],
+    xgb_runs[[3]]$evaluation_log[xgb_runs[[3]]$best_iteration, ]
+  ) %>%
+  cbind(model = c("mdl_max_bin_10", "mdl_max_bin_20", "mdl_max_bin_50"))
+
+  xgb_best <- xgb_runs_results$model[which.max(xgb_runs_results$test_auc)]
+
+  xgb_selected <- if (xgb_best == "mdl_max_bin_50") {
+    xgb_runs[[3]]
+  } else if (xgb_best == "mdl_max_bin_20") {
+    xgb_runs[[2]]
+  } else {
+    xgb_runs[[1]]
+  }
+
+  xgboost::xgb.model.dt.tree(
```

Comment:
Maybe return a named list of vectors of bin values (this looks like it is returning the tree object if I understand it).

Reply:
I actually like the returned format of this because it allows to clearly see the entire splitting path:

```
> xgboost::xgb.model.dt.tree(
+     model = xgb_mdl,
+     trees = xgb_mdl$best_iteration,
+     use_int_id = TRUE
+   )
   Tree Node Feature Split Yes No Missing      Quality     Cover
1:   11    0       0    19   1  2       1  5.491004470 505.31586
2:   11    1    Leaf    NA  NA NA      NA -0.001171782 467.08981
3:   11    2    Leaf    NA  NA NA      NA  0.115687110  38.22604
```
How about leaving it as is?

Reply:
It seems a lot to carry around though and adds complexity. Do you need all of the columns? Can you convert it to a tibble (since we keep everything in `recipes` as tibbles)?

Reply:
I think we could stick with Node, Feature, Split, Yes, No & Missing. I would keep these 6 columns for now especially for debugging purposes. When we settle down with this a little bit we could probably further limit them.